### DISCUSSION TITLE
Wrong evaluation of `sys.version_info` comparison

### DISCUSSION THREAD

**Discussion by Viicos** [microsoft/pyright#8648] [Q&A]
Consider the following example:

Code sample in [pyright playground](https://pyright-play.net/?pythonVersion=3.8&reportMissingModuleSource=true&code=JYWwDg9gTgLgBAZwJ4IFCuAM0SgdANwFMoFgIA7AfWHMwjgB4BeOACgGYAaOADgEoAXKjgi4AYjgAlQgEMAxgAsZAIwA2hOHIgATQsNEy4LAIypCqhISGjxcAKrkosxSvWadem4ZNA)

```python
import sys

if sys.version_info <= (3, 8):
    # Reachable code
    a = 1
else:
    # Unreachable code
    a = 1
```

When the Python version is set to `3.8`, pyright infers the first code path as being reachable. However, at runtime, `sys.version_info <= (3, 8)` evaluates to `False` (even when the micro version is `0`). Only when comparing `sys.version_info <= (3, 8, 0, "final", 0)` it will evaluate to `True`.

Is this expected behavior from pyright?

**Comment by erictraut** [comment]
When you say "set the Python version is set to `3.8`", I presume that you mean that you're specifying `"pythonVersion": "3.8"` manually in the configuration file. If you do not specify the python version in the configuration file, pyright will use the version from the configured Python environment, and this version will include a full version (including the micro version, release level, and serial number).

If you want to provide a more specific version in the config file, you can do so. For example, if you are targeting version "3.8.1" or "3.8.0.final.2", you can specify those versions in the config. However, if you specify only "3.8", pyright will assume wildcard values for the remaining unspecified components "3.8.?.?.?".

**Reply by Viicos** [reply]
> When you say "set the Python version is set to `3.8`", I presume that you mean that you're specifying `"pythonVersion": "3.8"` manually in the configuration file.

That's right, sorry for being unclear on that point.

The thing is when using `<=` to compare versions, `sys.version_info <= (3, 8)` will evaluate to `False` no matter the micro version of Python. That is, at runtime:

- `(sys.version_info <= (3, 8)) == False` with Python `3.8.0`
- `(sys.version_info <= (3, 8)) == False` with Python `3.8.x`, `x>=1`.

However, pyright evaluates it to `True`. I guess we are hitting a edge case here and anyway using `<=` isn't recommended (see [this Ruff rule](https://docs.astral.sh/ruff/rules/bad-version-info-comparison/) for example). So I don't know if it's worth changing pyright's behavior here.
